util/src/KubernetesClient.Util/Cache/Cache.cs

```diff
@@ -0,0 +1,429 @@
+    public Cache(string indexName, Func<TApiType, List<string>> indexFunc, Func<TApiType, string> keyFunc)
+    {
+      _indexers[indexName] = indexFunc;
+      _keyFunc = keyFunc;
+      _indices[indexName] = new Dictionary<string, HashSet<string>>();
+    }
+
+    /// <summary>
+    /// Add objects.
+    /// </summary>
+    /// <param name="obj">the obj</param>
+    public void Add(TApiType obj)
+    {
+      var key = _keyFunc(obj);
+
+      lock (_lock)
+      {
+        var oldObj = _items.GetValueOrDefault(key);
+        _items[key] = obj;
+        UpdateIndices(oldObj, obj, key);
+      }
+    }
+
+    /// <summary>
+    /// Update the object.
+    /// </summary>
+    /// <param name="obj">the obj</param>
+    public void Update(TApiType obj)
+    {
+      var key = _keyFunc(obj);
+
+      lock (_lock)
+      {
+        var oldObj = _items.GetValueOrDefault(key);
+        _items[key] = obj;
+        UpdateIndices(oldObj, obj, key);
+      }
+    }
+
+    /// <summary>
+    /// Delete the object.
+    /// </summary>
+    /// <param name="obj">the obj</param>
+    public void Delete(TApiType obj)
+    {
+      var key = _keyFunc(obj);
+      lock (_lock)
+      {
+        if (!_items.TryGetValue(key, out var value))
+        {
+          return;
+        }
+
+        DeleteFromIndices(value, key);
+        _items.Remove(key);
+      }
+    }
+
+    /// <summary>
+    /// Replace the content in the cache completely.
+    /// </summary>
+    /// <param name="list">the list</param>
+    /// <param name="resourceVersion">optional, unused param from interface</param>
+    /// <exception cref="ArgumentNullException">list is null</exception>
+    public void Replace(IEnumerable<TApiType> list, string resourceVersion = default)
+    {
+      if (list is null)
+      {
+        throw new ArgumentNullException(nameof(list));
+      }
+
+      lock (_lock)
+      {
+        var newItems = new Dictionary<string, TApiType>();
+        foreach (var item in list)
+        {
+          var key = _keyFunc(item);
+          newItems[key] = item;
+        }
+
+        _items = newItems;
+
+        // rebuild any index
+        _indices = new Dictionary<string, Dictionary<string, HashSet<string>>>();
+        foreach (var (key, value) in _items)
+        {
+          UpdateIndices(default, value, key);
+        }
+      }
+    }
+
+    /// <summary>
+    /// Resync.
+    /// </summary>
+    public void Resync()
+    {
+      // Do nothing by default
+    }
+
+    /// <summary>
+    /// List keys.
+    /// </summary>
+    /// <returns>the list</returns>
+    public IEnumerable<string> ListKeys()
+    {
+      return _items.Select(item => item.Key);
+    }
+
+    /// <summary>
+    /// Get object t.
+    /// </summary>
+    /// <param name="obj">the obj</param>
+    /// <returns>the t</returns>
+    public TApiType Get(TApiType obj)
+    {
+      var key = _keyFunc(obj);
+
+      lock (_lock)
```

Comment:
Let's at least add a TODO to make this lock striped or reader/writer (or use `ConcurrentDictionary`) I'm worried about lock contention with lots of getters.